### PR TITLE
webapi fixes

### PR DESCRIPTION
**Pull request by momchil-flex**
Credential folder only created when needed
Maximum attempts in authentication while loops
Authentication only called when needed, and checks if already done

### CODE REVIEW

File: tidy3d/web/auth.py [hunk 1]
```diff
@@ -46,7 +47,13 @@ def encode_password(password: str) -> str:
 
 def get_credentials() -> None:
     """Tries to log user in from environment variables, then from file, if not working, prompts
-    user for login info and saves to file."""
+    user for login info and saves to file.
+
+    If Config["auth" is already set, returns immediately.
```
Comment (tylerflex) [inline tidy3d/web/auth.py:52]: typo

File: tidy3d/web/auth.py [hunk 2]
```diff
@@ -111,4 +132,11 @@ def get_credentials() -> None:
         print(f"Unknown response: {keep_logged_in}")
 
 
-# get_credentials()
+def requires_auth(func):
+    """Decorator for functions that require the authentication step."""
+
+    def auth_before_call(*args, **kwargs):
```
Comment (tylerflex) [inline tidy3d/web/auth.py:138]: add a functools.wraps here to copy the docstring info over

**Review by tylerflex — CHANGES_REQUESTED** [review]

**Comment by tylerflex** [comment]
So what is the new behavior of this in terms of users?  Will it try to log in each web api call?  

Also, does this print anything to STDOUT each auth call because that could be annoying eg. if `web.run()` we will get like 7 outputs?  I'm fine with reducing the print output in get_credentials.

Also, can we use log.info instead of print?

**Comment by lei-flex** [comment]
os.environ["TIDY3D_USER"]
os.environ["TIDY3D_PASS"]

**Comment by lei-flex** [comment]
when credential is verified, we can also set values back to env.
os.environ["TIDY3D_USER"]
os.environ["TIDY3D_PASS"]

also the current solution requested unnecessary calls for auth. It is fine for right now.
the correct way is check the response for each call, if 401 error, then retrieve the credential again for your refrence: 
def handle_response(func):
    def wrapper(*args, **kwargs):
        resp = func(*args, **kwargs)
        if resp.status_code == 401 and Config.auth_retry < 1:
            getEmailPasswdAuthKey()
            Config.auth_retry = Config.auth_retry + 1
            resp = func(*args, **kwargs)
        elif resp.status_code != 200:
            print(resp.content)
            resp.raise_for_status()
        try:
            Config.auth_retry = 0
            jsn = resp.json()['data']
        except Exception as e:
            print('Could not json decode response : {0}!'.format(resp.text))
            raise
        return jsn

    return wrapper

File: tidy3d/web/auth.py [hunk 3]
```diff
@@ -3,24 +3,30 @@
 import getpass
 import hashlib
 import json
+import logging
```
Comment (tylerflex) [inline tidy3d/web/auth.py:6]: `from ..log import log`

File: tidy3d/web/auth.py [hunk 4]
```diff
@@ -44,25 +50,32 @@ def encode_password(password: str) -> str:
...
     """Tries to log user in from environment variables, then from file, if not working, prompts
     user for login info and saves to file."""
 
     # if we find credentials in environment variables
-    if "TIDY3D_USER" in os.environ and "TIDY3D_PASS" in os.environ:
-        print("Using Tidy3D credentials from enviornment")
+    if "TIDY3D_USER" in os.environ and (
+        "TIDY3D_PASS" in os.environ or "TIDY3D_PASS_HASH" in os.environ
+    ):
+        logging.debug("Using Tidy3D credentials from enviornment")
```
Comment (tylerflex) [inline tidy3d/web/auth.py:62]: `log.debug`

File: tidy3d/web/auth.py [hunk 5]
```diff
@@ -44,25 +50,32 @@ def encode_password(password: str) -> str:
...
             return
 
         except Exception:  # pylint:disable=broad-except
-            print("Error: Failed to log in with environment credentials.")
+            logging.info("Error: Failed to log in with environment credentials.")
 
     # if we find something in the credential path
     if os.path.exists(credential_path):
-        print("Using Tidy3D credentials from stored file")
+        logging.info("Using Tidy3D credentials from stored file")
```
Comment (tylerflex) [inline tidy3d/web/auth.py:78]: `log.info`

File: tidy3d/web/auth.py [hunk 6]
```diff
@@ -44,25 +50,32 @@ def encode_password(password: str) -> str:
...
+        else:
+            password = encode_password(password)
         try:
-            set_authentication_config(email, encode_password(password))
+            set_authentication_config(email, password)
             return
 
         except Exception:  # pylint:disable=broad-except
-            print("Error: Failed to log in with environment credentials.")
+            logging.info("Error: Failed to log in with environment credentials.")
```
Comment (tylerflex) [inline tidy3d/web/auth.py:74]: `log.info`

File: tidy3d/web/auth.py [hunk 7]
```diff
@@ -73,10 +86,13 @@ def get_credentials() -> None:
             return
 
         except Exception:  # pylint:disable=broad-except
-            print("Error: Failed to log in with saved credentials.")
+            logging.info("Error: Failed to log in with saved credentials.")
```
Comment (tylerflex) [inline tidy3d/web/auth.py:89]: `log.info`

File: tidy3d/web/auth.py [hunk 8]
```diff
@@ -73,10 +86,13 @@ def get_credentials() -> None:
...
 
     # keep trying to log in
-    while True:
+    attempts = 0
```
Comment (tylerflex) [inline tidy3d/web/auth.py:92]: could also just do `for attempts in range(MAX_ATTEMPTS):` I think (or `for _ in range(MAX_ATTEMPTS)` if `attempts` is not used.

File: tidy3d/web/auth.py [hunk 9]
```diff
@@ -88,27 +104,46 @@ def get_credentials() -> None:
...
+                if not os.path.exists(credential_dir):
+                    os.mkdir(credential_dir)
+
+                auth_json = {"email": email, "password": password}
+                with open(credential_path, "w", encoding="utf-8") as fp:
+                    json.dump(auth_json, fp)
+                return
+
+            except Exception:  # pylint:disable=broad-except
+                logging.info("Error: Failed to store credentials.")
```
Comment (tylerflex) [inline tidy3d/web/auth.py:130]: `log.info`

File: tidy3d/web/auth.py [hunk 10]
```diff
@@ -88,27 +104,46 @@ def get_credentials() -> None:
             break
 
         except Exception:  # pylint:disable=broad-except
-            print("Error: Failed to log in with new username and password.")
+            logging.info("Error: Failed to log in with new username and password.")
```
Comment (tylerflex) [inline tidy3d/web/auth.py:107]: `log.info`

File: tidy3d/web/auth.py (hunk 9 above)
Comment (tylerflex) [inline tidy3d/web/auth.py:130]: `log.info`

**Pull request merged**